### PR TITLE
[Feature] add InterCLR

### PR DESCRIPTION
**Pull request by Jiahao000**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add InterCLR ([Delving into Inter-Image Invariance for Unsupervised Visual Representations](https://link.springer.com/article/10.1007/s11263-022-01681-x), IJCV'22) to MMSelfSup

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: docs/en/model_zoo.md [hunk 1]
```diff
@@ -23,6 +23,7 @@ All models and part of benchmark results are recorded below.
...
 | [MoCo v3](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/mocov3/README.md)                    | [mocov3_vit-small-p16_32xb128-fp16-coslr-300e_in1k-224](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/mocov3/mocov3_vit-small-p16_32xb128-fp16-coslr-300e_in1k-224.py) | [model](https://download.openmmlab.com/mmselfsup/moco/mocov3_vit-small-p16_32xb128-fp16-coslr-300e_in1k-224_20220225-e31238dd.pth) \| [log](https://download.openmmlab.com/mmselfsup/moco/mocov3_vit-small-p16_32xb128-fp16-coslr-300e_in1k-224_20220222_160222.log.json) |
+| [InterCLR](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/interclr/README.md)                 | [interclr-moco_resnet50_8xb32-coslr-200e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/interclr/interclr-moco_resnet50_8xb32-coslr-200e_in1k.py)                 | [model](https://download.openmmlab.com/mmselfsup/interclr/interclr-moco_resnet50_8xb32-coslr-200e_in1k_20221206-38f8fdaf.pth) \| [log](https://download.openmmlab.com/mmselfsup/interclr/interclr-moco_resnet50_8xb32-coslr-200e_in1k_20221103_085100.log.json)           |
```
Comment (fangyixiao18) [inline docs/en/model_zoo.md:26]: also add linear evaluation results in `ImageNet Linear Evaluation` table?

File: docs/zh_cn/model_zoo.md [hunk 2]
```diff
@@ -23,6 +23,7 @@
...
 | [MoCo v3](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/mocov3/README.md)                    | [mocov3_vit-small-p16_32xb128-fp16-coslr-300e_in1k-224](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/mocov3/mocov3_vit-small-p16_32xb128-fp16-coslr-300e_in1k-224.py) | [model](https://download.openmmlab.com/mmselfsup/moco/mocov3_vit-small-p16_32xb128-fp16-coslr-300e_in1k-224_20220225-e31238dd.pth) \| [log](https://download.openmmlab.com/mmselfsup/moco/mocov3_vit-small-p16_32xb128-fp16-coslr-300e_in1k-224_20220222_160222.log.json) |
+| [InterCLR](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/interclr/README.md)                 | [interclr-moco_resnet50_8xb32-coslr-200e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/interclr/interclr-moco_resnet50_8xb32-coslr-200e_in1k.py)                 | [model](https://download.openmmlab.com/mmselfsup/interclr/interclr-moco_resnet50_8xb32-coslr-200e_in1k_20221206-38f8fdaf.pth) \| [log](https://download.openmmlab.com/mmselfsup/interclr/interclr-moco_resnet50_8xb32-coslr-200e_in1k_20221103_085100.log.json)           |
```
Comment (fangyixiao18) [inline docs/zh_cn/model_zoo.md:26]: the same comments with en version

File: mmselfsup/models/algorithms/interclr_moco.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+        ptr = int(self.queue_ptr)
+        assert self.queue_len % batch_size == 0  # for simplicity
+
+        # replace the keys at ptr (dequeue and enqueue)
+        self.queue[:, ptr:ptr + batch_size] = keys.transpose(0, 1)
+        ptr = (ptr + batch_size) % self.queue_len  # move pointer
+
+        self.queue_ptr[0] = ptr
+
+    def contrast_intra(self, q, k):
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/interclr_moco.py:180]: can this function be reused in InterCLR-NPID or  InterCLR-BYOL?

File: mmselfsup/models/algorithms/interclr_moco.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+                    cosine > math.cos(math.pi - self.intra_arc_margin), phi,
+                    cosine - math.sin(math.pi - self.intra_arc_margin) *
+                    self.intra_arc_margin)
+            pos_logits.copy_(phi)
+
+        losses = self.head(pos_logits, neg_logits)
+
+        return losses
+
+    def contrast_inter(self, q, idx):
```
Comment (fangyixiao18) [inline mmselfsup/models/algorithms/interclr_moco.py:227]: the same question as function `contrast_intra`, besides better to add more docstrings to explain these two functions

File: mmselfsup/models/algorithms/interclr_moco.py (hunk 3 above)
Comment (Jiahao000) [reply mmselfsup/models/algorithms/interclr_moco.py:180]: It cannot be reused as different methods have different forms of loss.

File: mmselfsup/models/algorithms/interclr_moco.py (hunk 4 above)
Comment (Jiahao000) [reply mmselfsup/models/algorithms/interclr_moco.py:227]: same as `contrast_intra`

File: configs/selfsup/interclr/README.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+The **Feature1 - Feature5** don't have the GlobalAveragePooling, the feature map is pooled to the specific dimensions and then follows a Linear layer to do the classification. Please refer to [resnet50_mhead_linear-8xb32-steplr-90e_in1k](https://github.com/open-mmlab/mmselfsup/blob/master/configs/benchmarks/classification/imagenet/resnet50_mhead_linear-8xb32-steplr-90e_in1k.py) for details of config.
+
+| Self-Supervised Config                                                                                                                                                  | Feature1 | Feature2 | Feature3 | Feature4 | Feature5 |
+| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------- | -------- | -------- | -------- | -------- |
+| [interclr-moco_resnet50_8xb32-coslr-200e](https://github.com/open-mmlab/mmselfsup/blob/master/configs/selfsup/interclr/interclr-moco_resnet50_8xb32-coslr-200e_in1k.py) | 15.59    | 35.10    | 47.36    | 62.86    | 68.04    |
+
+## Citation
+
+```bibtex
```
Comment (YuanLiuuuuuu) [inline configs/selfsup/interclr/README.md:55]: The `bibtex` can be reformatted. You can refer to the style of MoCo.

File: mmselfsup/models/memories/interclr_memory.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        length (int): Number of features stored in the memory bank.
+        feat_dim (int): Dimension of stored features.
+        momentum (float): Momentum coefficient for updating features.
+        num_classes (int): Number of clusters.
+        min_cluster (int): Minimal cluster size.
+    """
+
+    def __init__(self, length, feat_dim, momentum, num_classes, min_cluster,
+                 **kwargs):
+        super(InterCLRMemory, self).__init__()
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/memories/interclr_memory.py:27]:
```suggestion
        super().__init__()
```

File: mmselfsup/models/memories/interclr_memory.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        assert len(max_cluster_inds) >= 2
+        max_cluster_features = self.feature_bank[
+            max_cluster_inds, :].cpu().numpy()
+        if np.any(np.isnan(max_cluster_features)):
+            raise Exception('Has nan in features.')
+        kmeans_ret = self.kmeans.fit(max_cluster_features)
+        sub_cluster1_ind = max_cluster_inds[kmeans_ret.labels_ == 0]
+        sub_cluster2_ind = max_cluster_inds[kmeans_ret.labels_ == 1]
+        if not (len(sub_cluster1_ind) > 0 and len(sub_cluster2_ind) > 0):
+            print(
```
Comment (YuanLiuuuuuu) [inline mmselfsup/models/memories/interclr_memory.py:207]: better use ``logging`` instead of ``print``

File: mmselfsup/models/memories/interclr_memory.py (hunk 6 above)
Comment (Jiahao000) [reply mmselfsup/models/memories/interclr_memory.py:27]: This follows the same style in `ODCMemory` and `SimpleMemory`

File: mmselfsup/models/memories/interclr_memory.py (hunk 7 above)
Comment (Jiahao000) [reply mmselfsup/models/memories/interclr_memory.py:207]: This is the same as `ODCMemory`

**Review by fangyixiao18 — APPROVED** [review]
LGTM.

**Review by YuanLiuuuuuu — APPROVED** [review]
